ext/nokogiri/xml_namespace.c

```diff
@@ -34,21 +62,37 @@ static VALUE href(VALUE self)
   return NOKOGIRI_STR_NEW2(ns->href);
 }
 
+static int part_of_an_xpath_node_set_eh(xmlNsPtr node)
+{
+  return (node->next && ! Nokogiri_namespace_eh(node->next));
```

Comment:
See https://github.com/GNOME/libxml2/blob/master/xpath.c#L4190-L4191 for original source. Also check out the comments I made in xml_node_set.c in this series of patches.

For a better explanation, I'd recommend reading the patches in order, I tried to tell a story with them.